### DISCUSSION TITLE
Cannot use `useContractWrite` with zkSync-era Paymaster

### DISCUSSION THREAD

**Discussion by equalogix** [wevm/wagmi#2210] [Q&A]
When using `useContractWrite` to call a smart contract function using the zkSync-era paymaster I get the following error:
```
Error: invalid object key - customData (argument="transaction:customData", value={"data":"0x9397badb0000...", "from": ...etc... }
```

Here is the code that makes the call:
```
const { config, error } = usePrepareContractWrite({
    address: CONTRACT_ADDRESS,
    abi: CONTRACT_ABI,
    functionName: 'mintToken',
    args: [hash, uri, itemId],
    overrides: {
      customData: {
        gasPerPubdata: 50000,
        paymasterParams: paymasterParams,
    }},
  });
const { data, isLoading, isSuccess, write } = useContractWrite(config);
// note that paymasterParams is the return value from getPaymasterParams() 
//JSX
<Button disabled={!write} onClick={() => write?.()}>Mint</Button>
```
I read through the docs and I don't see any mention for support of `customData`. I also looked through the viem docs (for future use) and again no mention. In fact in the viem case, there is a the switch from overrides object to explicitly calling the overrides attributes, but there is no attribute for customData.

**Comment by equalogix** [comment]
I also tried using the `mode: 'recklesslyUnprepared',` and it resulted in the same error as above.

**Comment by caffeinum** [comment]
Have the same issue, can't figure out how to do it

**Comment by caffeinum** [comment]
in fact, `customData` is zksync-specific field and it is not present in original `ethers` [1], hence probably `wagmi` will not have any support for it until they integrate `zksync-web3` into the library

[1] https://era.zksync.io/docs/api/js/features.html#overrides

**Comment by equalogix** [comment]
I have been able to get it to work on desktop using both `Walletconnect` and `Metamask/SDK` but in both cases I am not using Wagmi for the contract interaction, I use `ethers.js` using `sendTransaction` to interact with the contracts. `WalletConnect` doesn't work on Mobile (web browser) on either iOS or Android where as `Metamask/SDK` works on both (iOS/Android). But I have trouble connecting the wallet with `Metamask/SDK` on Android. So as it stands, as far as I know there is no production ready solution for this.

It is also worth pointing out that Wagmi is now using `viem` but I don't believe it has anymore support for `customData` then `ether.js`.

**Comment by dutterbutter** [comment]
Hey @equalogix @caffeinum, we are actively looking to get paymaster support in `viem` which would unblock the usage in `wagmi`. We opened a discussion post to get feedback from the maintainers there. Feel free to check it out [here](https://github.com/wagmi-dev/viem/discussions/1229) and post any relevant commentary.
